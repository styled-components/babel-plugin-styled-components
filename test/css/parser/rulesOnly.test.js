import parse from '../../../src/css/parser';

describe('parse rules only', () => {
  it('parses normal rules', () => {
    const ast = parse`
      .class {
      }
    `

    expect(ast).toMatchSnapshot();
    expect(ast.rules[0].selector.value.trim()).toBe('.class');
  });

  it('parses multiple rules', () => {
    const ast = parse`
      .classA {}
      .classB {}
    `

    expect(ast).toMatchSnapshot();
    expect(ast.rules[0].selector.value.trim()).toBe('.classA');
    expect(ast.rules[1].selector.value.trim()).toBe('.classB');
  });

  it('parses nested rules', () => {
    const ast = parse`
      .outer {
        .inner {}
      }
    `

    expect(ast).toMatchSnapshot();
    expect(ast.rules[0].selector.value.trim()).toBe('.outer');
  });

  it('parses multiple nested rules', () => {
    const ast = parse`
      .outerA {
        .innerA {}
      }

      .outerB {
        .innerB {}
      }
    `

    expect(ast).toMatchSnapshot();
    expect(ast.rules[0].selector.value.trim()).toBe('.outerA');
    expect(ast.rules[1].selector.value.trim()).toBe('.outerB');
  });

  it('parses rules containing declarations', () => {
    const ast = parse`
      .class {
        text-align: center;
      }
    `

    expect(ast).toMatchSnapshot();
    expect(ast.rules[0].block.declarations[0].property.value.trim()).toBe('text-align');
    expect(ast.rules[0].block.declarations[0].value.value.trim()).toBe('center');
  });

  it('throws if closing paranthesis is missing', () => {
    expect(() => parse`
      .class {
        text-align: center;
    `).toThrow()
  });

  it('throws if closing paranthesis is missing', () => {
    expect(() => parse(`
      .classA {
        text-align: center;
        .classB {}
    `)).toThrow()
  });
});
