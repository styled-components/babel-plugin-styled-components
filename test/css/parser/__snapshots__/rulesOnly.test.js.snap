exports[`parse rules only parses multiple nested rules 1`] = `
Object {
  "declarations": Array [],
  "loc": Array [
    0,
    91,
  ],
  "rules": Array [
    Object {
      "block": Object {
        "declarations": Array [],
        "loc": Array [
          16,
          42,
        ],
        "rules": Array [
          Object {
            "block": Object {
              "declarations": Array [],
              "loc": Array [
                34,
                34,
              ],
              "rules": Array [],
              "type": "Block",
            },
            "loc": Array [
              16,
              35,
            ],
            "selector": Object {
              "interpolations": Array [],
              "loc": Array [
                16,
                32,
              ],
              "type": "SelectorLiteral",
              "value": "
        .innerA",
            },
            "type": "Rule",
          },
        ],
        "type": "Block",
      },
      "loc": Array [
        0,
        43,
      ],
      "selector": Object {
        "interpolations": Array [],
        "loc": Array [
          0,
          14,
        ],
        "type": "SelectorLiteral",
        "value": "
      .outerA",
      },
      "type": "Rule",
    },
    Object {
      "block": Object {
        "declarations": Array [],
        "loc": Array [
          60,
          86,
        ],
        "rules": Array [
          Object {
            "block": Object {
              "declarations": Array [],
              "loc": Array [
                78,
                78,
              ],
              "rules": Array [],
              "type": "Block",
            },
            "loc": Array [
              60,
              79,
            ],
            "selector": Object {
              "interpolations": Array [],
              "loc": Array [
                60,
                76,
              ],
              "type": "SelectorLiteral",
              "value": "
        .innerB",
            },
            "type": "Rule",
          },
        ],
        "type": "Block",
      },
      "loc": Array [
        44,
        87,
      ],
      "selector": Object {
        "interpolations": Array [],
        "loc": Array [
          44,
          58,
        ],
        "type": "SelectorLiteral",
        "value": "
      .outerB",
      },
      "type": "Rule",
    },
  ],
  "type": "Block",
}
`;

exports[`parse rules only parses multiple rules 1`] = `
Object {
  "declarations": Array [],
  "loc": Array [
    0,
    38,
  ],
  "rules": Array [
    Object {
      "block": Object {
        "declarations": Array [],
        "loc": Array [
          16,
          16,
        ],
        "rules": Array [],
        "type": "Block",
      },
      "loc": Array [
        0,
        17,
      ],
      "selector": Object {
        "interpolations": Array [],
        "loc": Array [
          0,
          14,
        ],
        "type": "SelectorLiteral",
        "value": "
      .classA",
      },
      "type": "Rule",
    },
    Object {
      "block": Object {
        "declarations": Array [],
        "loc": Array [
          33,
          33,
        ],
        "rules": Array [],
        "type": "Block",
      },
      "loc": Array [
        18,
        34,
      ],
      "selector": Object {
        "interpolations": Array [],
        "loc": Array [
          18,
          31,
        ],
        "type": "SelectorLiteral",
        "value": "      .classB",
      },
      "type": "Rule",
    },
  ],
  "type": "Block",
}
`;

exports[`parse rules only parses nested rules 1`] = `
Object {
  "declarations": Array [],
  "loc": Array [
    0,
    45,
  ],
  "rules": Array [
    Object {
      "block": Object {
        "declarations": Array [],
        "loc": Array [
          15,
          40,
        ],
        "rules": Array [
          Object {
            "block": Object {
              "declarations": Array [],
              "loc": Array [
                32,
                32,
              ],
              "rules": Array [],
              "type": "Block",
            },
            "loc": Array [
              15,
              33,
            ],
            "selector": Object {
              "interpolations": Array [],
              "loc": Array [
                15,
                30,
              ],
              "type": "SelectorLiteral",
              "value": "
        .inner",
            },
            "type": "Rule",
          },
        ],
        "type": "Block",
      },
      "loc": Array [
        0,
        41,
      ],
      "selector": Object {
        "interpolations": Array [],
        "loc": Array [
          0,
          13,
        ],
        "type": "SelectorLiteral",
        "value": "
      .outer",
      },
      "type": "Rule",
    },
  ],
  "type": "Block",
}
`;

exports[`parse rules only parses normal rules 1`] = `
Object {
  "declarations": Array [],
  "loc": Array [
    0,
    27,
  ],
  "rules": Array [
    Object {
      "block": Object {
        "declarations": Array [],
        "loc": Array [
          15,
          22,
        ],
        "rules": Array [],
        "type": "Block",
      },
      "loc": Array [
        0,
        23,
      ],
      "selector": Object {
        "interpolations": Array [],
        "loc": Array [
          0,
          13,
        ],
        "type": "SelectorLiteral",
        "value": "
      .class",
      },
      "type": "Rule",
    },
  ],
  "type": "Block",
}
`;

exports[`parse rules only parses rules containing declarations 1`] = `
Object {
  "declarations": Array [],
  "loc": Array [
    0,
    55,
  ],
  "rules": Array [
    Object {
      "block": Object {
        "declarations": Array [
          Object {
            "loc": Array [
              15,
              42,
            ],
            "property": Object {
              "loc": Array [
                15,
                34,
              ],
              "type": "PropertyLiteral",
              "value": "
        text-align",
            },
            "type": "Declaration",
            "value": Object {
              "interpolations": Array [],
              "loc": Array [
                35,
                42,
              ],
              "type": "ValueLiteral",
              "value": " center",
            },
          },
        ],
        "loc": Array [
          15,
          50,
        ],
        "rules": Array [],
        "type": "Block",
      },
      "loc": Array [
        0,
        51,
      ],
      "selector": Object {
        "interpolations": Array [],
        "loc": Array [
          0,
          13,
        ],
        "type": "SelectorLiteral",
        "value": "
      .class",
      },
      "type": "Rule",
    },
  ],
  "type": "Block",
}
`;
